### PR TITLE
feat(sink): upsert avro with schema registry

### PR DESCRIPTION
**Pull request by xiangjinwu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

* Only upsert, not append-only yet
  * Shorter to review. Will refactor to make it less repetitive.
* Only avro, not protobuf yet
  * Requires more header ([doc](https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/index.html#messages-wire-format))
* Only schema registry, not file / https / s3 yet
  * #12871
* Only kafka, not kinesis or pulsar
  * Kinesis / Plusar does not allow key to be non-string bytes
  * Pulsar has its own schema registry ([doc](https://pulsar.apache.org/docs/3.1.x/schema-overview/)).

Note that apeend-only protobuf with file / https in kafka has been supported by #12858. We will fill the support matrix gradually.

`SinkFormatterImpl::new` will be refactored in a followup.
* `topic` is used in schema registry name strategy, but does not apply to Redis.
* `db_name` and `sink_from_name` is only useful for kafka connect (incl debezium) metadata.
* `AwsAuthProps` is only useful when loading schema definition from s3.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Supports kafka sink with upsert avro using schema registry.

Example `create sink`  command can be found in `e2e_test/sink/kafka/avro.slt`

Same as source, the following options can be used after `format upsert encode avro`:
* `schema.registry`
* `schema.registry.username`
* `schema.registry.password`
* `schema.registry.name.strategy` (optional)
* `key.message` (only when name strategy is set to `record_name_strategy` or `topic_record_name_strategy`)
* `message` (only when name strategy is set to `record_name_strategy` or `topic_record_name_strategy`)

### CODE REVIEW

**Comment by xiangjinwu** [comment]
Conflicts resolved and ready for review

File: src/connector/src/schema/avro.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            id: fetched.id,
+        })
+    }
+}
+
+/// Schema registry only
+pub async fn fetch_schema(
+    format_options: &BTreeMap<String, String>,
+    topic: &str,
+) -> Result<(SchemaWithId, SchemaWithId), SchemaFetchError> {
```
Comment (tabVersion) [inline src/connector/src/schema/avro.rs:52]: copy from the source part?

File: src/connector/src/sink/encoder/avro.rs [hunk 2]
```diff
@@ -59,12 +81,19 @@ impl AvroEncoder {
             schema,
             col_indices,
             avro_schema,
+            header,
         })
     }
 }
 
+pub struct AvroEncoded {
```
Comment (tabVersion) [inline src/connector/src/sink/encoder/avro.rs:89]: @liurenjie1024 please check if we use schema registry correctly?

File: src/connector/src/sink/formatter/mod.rs [hunk 3]
```diff
@@ -225,6 +249,22 @@ macro_rules! dispatch_sink_formatter_impl {
             SinkFormatterImpl::AppendOnlyJson($name) => $body,
             SinkFormatterImpl::AppendOnlyProto($name) => $body,
             SinkFormatterImpl::UpsertJson($name) => $body,
+            SinkFormatterImpl::UpsertAvro($name) => $body,
+            SinkFormatterImpl::DebeziumJson($name) => $body,
+            SinkFormatterImpl::AppendOnlyTemplate($name) => $body,
+            SinkFormatterImpl::UpsertTemplate($name) => $body,
```
Comment (tabVersion) [inline src/connector/src/sink/formatter/mod.rs:255]: what template mean here?

**Review by tabVersion — APPROVED** [review]
generally LGTM

File: src/connector/src/sink/encoder/avro.rs (hunk 2 above)
Comment (liurenjie1024) [reply src/connector/src/sink/encoder/avro.rs:89]: LGTM

File: src/connector/src/schema/avro.rs (hunk 1 above)
Comment (xiangjinwu) [reply src/connector/src/schema/avro.rs:52]: They will be merged eventually.

File: src/connector/src/sink/formatter/mod.rs (hunk 3 above)
Comment (xiangjinwu) [reply src/connector/src/sink/formatter/mod.rs:255]: `format upsert encode template (key_format = 'V1 is: {v1}', value_format = 'V2 is: {v2}')` initially introduced by redis sink.

**Pull request merged**